Add validation to bot tokens based on string length

Chris-Johnston opened the pull request:
Before generating the authorization header to the prefixed token, this change will check the length of bot tokens to see if they are at least 50 characters long. This is longer than the length of a bearer token or oauth client secret, so it throws an exception if the user tries to use the wrong token here.

I've checked against several of my tokens of various ages, and they all appear to be 59 characters long. [Even this example on the Discord api docs is 59 characters long.](https://discordapp.com/developers/docs/reference#authentication-example-bot-token-authorization-header) But, because this value isn't explicitly stated to always be of this length, I don't think that's a safe assumption to make.

Would it make sense to add this validation for all of the token types?

Joe4evr commented:
There's no guarantee from Discord that tokens will *stay* inside some range of length, far as I know.

But if we *are* going with this, aren't client secrets way longer than the auth token?

Chris-Johnston commented:
> aren't client secrets way longer than the auth token?

[Going off of what I can see from the docs](https://discordapp.com/developers/docs/topics/oauth2#authorization-code-grant-access-token-exchange-example) and my own applications, client secrets look to be significantly shorter. Discord's example is 25 characters long, mine was 32. So it looks like the bot tokens are longer.

Commit:
Add input validation for bot tokens based on their length

Commit:
Add token validation to BaseDiscordClient#LoginAsync

Adds a TokenUtils class which is used to validate that tokens are correct

Commit:
Revert changes to DiscordRestApiClient

Commit:
Add Unit tests to the TokenUtils class, fix a logic error that was caught by those tests

Joe4evr commented:
LGTM, I guess.

AntiTcb commented:
Was brought up earlier in the channel, it might be better to log a warning to the user and let the API throw the error.

Commit:
Allow for API to throw exceptions

Moves the validation of tokens to be inside of LoginInternalAsync, and writes a Warning to the console when the supplied tokens are invalid
